internal/impl/ollama/base_processor.go

```diff
@@ -0,0 +1,189 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Licensed as a Redpanda Enterprise file under the Redpanda Community
+// License (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+// https://github.com/redpanda-data/connect/blob/main/licenses/rcl.md
+
+package ollama
+
+import (
+	"context"
+	"errors"
+	"io"
+	"io/fs"
+	"net/http"
+	"net/url"
+	"os"
+	"os/exec"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/dustin/go-humanize"
+	"github.com/ollama/ollama/api"
+	"github.com/redpanda-data/benthos/v4/public/service"
+)
+
+const (
+	bopFieldServerAddress = "server_address"
+	bopFieldModel         = "model"
+)
+
+func extractEmbeddedServer(fsys *service.FS, serverPath string) error {
+	_, err := fsys.Stat(serverPath)
+	if err == nil {
+		return nil
+	}
+	if errors.Is(err, fs.ErrNotExist) {
+		file, err := fsys.OpenFile(serverPath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, os.ModePerm)
+		if err != nil {
+			return err
+		}
+		writer, isw := file.(io.Writer)
+		if !isw {
+			return errors.New("failed to write binary")
+		}
+		if _, err = writer.Write(embeddedServer); err != nil {
+			return err
+		}
+		return file.Close()
+	}
+	return err
+}
+
+func startServer(fs *service.FS) (*exec.Cmd, error) {
+	serverPath, err := exec.LookPath("ollama")
+	if errors.Is(err, exec.ErrNotFound) {
+		if embeddedServer == nil {
+			return nil, errors.New("embedded ollama not supported for this platform - please specify a server_address")
+		}
+		serverPath = path.Join(os.TempDir(), "ollama")
+		if err = extractEmbeddedServer(fs, serverPath); err != nil {
+			return nil, err
+		}
+	} else if err != nil {
+		return nil, err
+	}
+	proc := exec.Command(serverPath, "serve")
+	// proc.Stdout = os.Stdout
+	// proc.Stderr = os.Stderr
```

Comment:
If you have ideas on how to hook it up I'm all for doing it - it's actually pretty helpful for troubleshooting.

Reply:
Both `proc.Stderr` and `proc.Stdout` are of type [`io.Writer`](https://pkg.go.dev/io#Writer). I'd create a struct which holds a reference to the logger and implements this interface (it just needs a `Write(p []byte) (n int, err error)` method which does a write to the logger member variable from the receiver if that makes sense). Let me know if you need an example.

Reply:
Indeed, it might be more complicated. My initial thinking was that wazero is smart enough to do a single write / log entry, so, even if it contains new lines, we could emit those as-is and leave these as debug logs or, we could also escape the newline characters so we don't end up with multiline logs (I think there's some other stuff which emits multiline logs in debug mode).

Reply:
Took a stab at this 😄

Reply:
Awesome, thanks! That looks fine as long as there aren't any concurrent calls to `Write`. Might be worth protecting it with a mutex just in case.